crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -55,6 +59,11 @@ def checks
             passed: clusters_healthy?,
             errors: clusters_health_check_errors
           } if Api::Crowbar.addons.include?("ha")
+
+          return ret unless upgrade_status.current_step == "upgrade_prechecks"
+
+          errors = ret.select { |_k, v| v[:required] && v[:errors].any? }.map { |_k, v| v[:errors] }
+          upgrade_status.end_step(errors.any?, errors)
```

Comment:
Just occurred to me that we should make the state library defend against one thread starting a step and other ending it.  This should never happen unless the caller is really buggy, but it's best to be safe.  However that is not related to this PR at all.